service/metadata/service/permission.go

```diff
@@ -0,0 +1,260 @@
+		return nil, errors.ErrInvalidParams
+	}
+
+	operator, err = sdk.AccAddressFromHexUnsafe(req.Operator)
+	if err != nil {
+		log.CtxErrorw(ctx, "failed to creates an AccAddress from a HEX-encoded string", "error", err)
+		return nil, err
+	}
+
+	if req.BucketName == "" {
+		log.Errorw("failed to check bucket name", "bucket_name", req.BucketName, "error", err)
+		return nil, errors.ErrInvalidBucketName
+	}
+
+	bucketInfo, err = metadata.bsDB.GetBucketByName(req.BucketName, true)
+	if err != nil || bucketInfo == nil {
+		log.CtxErrorw(ctx, "failed to get bucket info", "error", err)
+		return nil, errors.ErrNoSuchBucket
+	}
+
+	if req.ObjectName == "" {
+		effect, err = metadata.VerifyBucketPermission(ctx, bucketInfo, operator, req.ActionType, nil)
+		if err != nil {
+			log.CtxErrorw(ctx, "failed to verify bucket permission", "error", err)
+			return nil, err
+		}
+	} else {
+		objectInfo, err = metadata.bsDB.GetObjectInfo(req.BucketName, req.ObjectName)
+		if err != nil || objectInfo == nil {
+			log.CtxErrorw(ctx, "failed to get object info", "error", err)
+			return nil, errors.ErrNoSuchObject
+		}
+		effect = metadata.VerifyObjectPermission(ctx, bucketInfo, objectInfo, operator, req.ActionType)
+	}
+
+	resp = &storagetypes.QueryVerifyPermissionResponse{Effect: effect}
+	log.CtxInfow(ctx, "succeed to get payment by bucket id")
+	return resp, nil
+}
+
+// VerifyBucketPermission verify bucket permission
+func (metadata *Metadata) VerifyBucketPermission(ctx context.Context, bucketInfo *bsdb.Bucket, operator sdk.AccAddress,
+	action permtypes.ActionType, options *permtypes.VerifyOptions) (permtypes.Effect, error) {
+	var (
+		err    error
+		owner  sdk.AccAddress
+		effect permtypes.Effect
+	)
+
+	// if bucket is public, anyone can read but can not write it.
+	if bucketInfo.Visibility == storagetypes.VISIBILITY_TYPE_PUBLIC_READ.String() && keeper.PublicReadBucketAllowedActions[action] {
+		return permtypes.EFFECT_ALLOW, nil
+	}
+
+	owner, err = sdk.AccAddressFromHexUnsafe(bucketInfo.Owner.String())
+	if err != nil {
+		log.CtxErrorw(ctx, "failed to creates an AccAddress from a HEX-encoded string", "error", err)
+		return permtypes.EFFECT_DENY, err
+	}
+
+	// the owner has full permissions
+	if operator.Equals(owner) {
+		return permtypes.EFFECT_ALLOW, nil
+	}
+
+	// verify policy
+	effect, err = metadata.VerifyPolicy(ctx, math.NewUintFromBigInt(bucketInfo.BucketID.Big()), gnfdresource.RESOURCE_TYPE_BUCKET, operator, action, options)
+	if err != nil {
+		log.CtxErrorw(ctx, "failed to verify bucket policy", "error", err)
+		return permtypes.EFFECT_DENY, err
+	}
+
+	if effect == permtypes.EFFECT_ALLOW {
+		return permtypes.EFFECT_ALLOW, nil
+	}
+	return permtypes.EFFECT_DENY, nil
+}
+
+// VerifyObjectPermission verify object permission
+func (metadata *Metadata) VerifyObjectPermission(ctx context.Context, bucketInfo *bsdb.Bucket, objectInfo *bsdb.Object,
+	operator sdk.AccAddress, action permtypes.ActionType) permtypes.Effect {
```

Comment:
This function does not return an error value, but VerifyBucketPermission returns.

Reply:
fixed, return an error value for VerifyObjectPermission